src/lib/tools/repeat_finder.rs

```diff
@@ -20,33 +29,141 @@ pub struct Opts {
 
     /// The minimum number of bases for a repeat region must span to be called a microsatellite
     #[clap(short = 'm', default_value = "10", long, display_order = 3)]
-    pub min_bases: u64,
+    pub min_bases: usize,
 
     /// The maximum number of bases for a repeat region must span to be called a microsatellite
     #[clap(short = 'M', default_value = "100", long, display_order = 4)]
-    pub max_bases: u64,
+    pub max_bases: usize,
 
     /// The minimum number of repeats for a repeat region must span to be called a microsatellite
     #[clap(short = 'r', default_value = "3", long, display_order = 5)]
-    pub min_repeats: u64,
+    pub min_repeats: usize,
 
     /// The minimum k-mer length
     #[clap(short = 'l', default_value = "1", long, display_order = 6)]
-    pub min_repeat_length: u64,
+    pub min_repeat_length: usize,
 
     /// The maximum k-mer length
     #[clap(short = 'L', default_value = "5", long, display_order = 7)]
-    pub max_repeat_length: u64,
+    pub max_repeat_length: usize,
 }
 
 // Run extract
 #[allow(clippy::too_many_lines)]
 pub fn run(opts: &Opts) -> Result<(), anyhow::Error> {
-    println!("{:?}", opts);
-    // TODO
+    // Build the finders
+    let mut finders: Vec<KmerFinder> = Vec::new();
+    let mut i = opts.min_repeat_length;
+    while i <= opts.max_repeat_length {
+        finders.push(KmerFinder::new(i));
+        i += 1;
+    }
+
+    // Open the input and output
+    let mut reader = File::open(&opts.input).map(BufReader::new).map(fasta::Reader::new)?;
+    let mut writer = File::create(&opts.output).map(BufWriter::new).map(bed::Writer::new)?;
+
+    // Go through each contig, one at a time
+    for result in reader.records() {
+        let record = result?;
+        let contig = record.name();
+
+        for finder in &mut finders {
+            finder.reset();
+        }
+
+        // Go through each base, one at a time
+        let mut i = 1;
+        while i <= record.sequence().len() {
+            let position = Position::try_from(i)?;
+            let base: u8 = record.sequence().get(position).unwrap() & 0xdf; // to upper case
+
+            // Output the smallest repeat found ending at this position.
+            let mut found = false;
+            for finder in &mut finders {
+                if let Some(repeat) = finder.add_maybe_emit(base, !found) {
+                    if found {
+                        continue;
+                    } else if let Some(rec) = to_bed_record(opts, contig, i, finder, &repeat) {
+                        writer.write_record(&rec).unwrap();
```

Comment:
If we're going to use unwraps I'd argue we should have some form of informative message associated with them, though I think Rust best practices would be to define custom Errors with `anyhow`, use `map_err` to convert the errors from `noodles` / etc into our errors, and then raise the error with `?`.